File: typescript/modules/libs/workspace/derivations/apps/src/lib/rules/part-category-display-name.rule.ts
import { Dependency, RoleType } from '@allors/workspace/meta/system';
import { IRule } from '@allors/workspace/domain/system';
import { M } from '@allors/workspace/meta/default';
import { PartCategory } from '@allors/workspace/domain/default';

export class PartCategoryDisplayNameRule implements IRule<PartCategory> {
  roleType: RoleType;
  dependencies: Dependency[];

  constructor(m: M) {
    const { treeBuilder: t, dependency: d } = m;

    this.roleType = m.PartCategory.DisplayName;

    this.dependencies = [d(m.PartCategory, (v) => v.PrimaryParent)];
  }

  derive(partCategory: PartCategory) {
    const selfAndPrimaryAncestors = [partCategory];
    let ancestor: PartCategory | null = partCategory;
    while (ancestor != null && selfAndPrimaryAncestors.indexOf(ancestor) < 0) {
      selfAndPrimaryAncestors.push(ancestor);
      ancestor = ancestor.PrimaryParent;
    }

    selfAndPrimaryAncestors.reverse();
    partCategory.DisplayName = selfAndPrimaryAncestors.map((v) => v.Name).join('/');
  }
}

File: dotnet/Apps/Database/Domain/Apps/Rules/Product/PartCategoryDisplayNameRule.cs
// <copyright file="NonUnifiedPartDerivation.cs" company="Allors bv">
// Copyright (c) Allors bv. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Database.Domain
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Database.Derivations;
    using Meta;
    using Derivations.Rules;

    public class PartCategoryDisplayNameRule : Rule
    {
        public PartCategoryDisplayNameRule(MetaPopulation m) : base(m, new Guid("ef3d0304-019d-4580-adae-3f5736ca0211")) =>
            this.Patterns = new Pattern[]
            {
                m.PartCategory.RolePattern(v => v.Name),
                m.PartCategory.RolePattern(v => v.PrimaryAncestors),
            };

        public override void Derive(ICycle cycle, IEnumerable<IObject> matches)
        {
            var validation = cycle.Validation;

            foreach (var @this in matches.Cast<PartCategory>())
            {
                @this.DerivePartCategoryDisplayName(validation);
            }
        }
    }

    public static class PartCategoryDisplayNameRuleExtensions
    {
        public static void DerivePartCategoryDisplayName(this PartCategory @this, IValidation validation)
        {
            var primaryAncestors = @this.PrimaryAncestors.Reverse().ToList();
            primaryAncestors.Add(@this);

            var array = new string[] {
                    string.Join(", ", string.Join("/", primaryAncestors.Select(v => v.Name ?? string.Empty).ToArray()))
                };

            if (array.Any(s => !string.IsNullOrEmpty(s)))
            {
                @this.DisplayName = string.Join(" ", array.Where(s => !string.IsNullOrEmpty(s)));
            }
            else
            {
                @this.RemoveDisplayName();
            }
        }
    }
}
